stl/inc/stop_token

```diff
@@ -0,0 +1,401 @@
+#else // ^^^ !_HAS_CXX20 / _HAS_CXX20 vvv
+
+#include <atomic>
+#include <xmemory>
+#include <xthreads.h>
+
+#pragma pack(push, _CRT_PACKING)
+#pragma warning(push, _STL_WARNING_LEVEL)
+#pragma warning(disable : _STL_DISABLED_WARNINGS)
+_STL_DISABLE_CLANG_WARNINGS
+#pragma push_macro("new")
+#undef new
+
+_STD_BEGIN
+struct nostopstate_t {
+    explicit nostopstate_t() = default;
+};
+
+inline constexpr nostopstate_t nostopstate{};
+
+struct _Stop_state;
+class stop_token;
+
+class _Stop_callback_base {
+    friend _Stop_state;
+
+public:
+    using _Callback_fn = void(__cdecl*)(_Stop_callback_base*) _NOEXCEPT_FNPTR;
+
+    explicit _Stop_callback_base(const _Callback_fn _Fn_) noexcept : _Fn{_Fn_} {}
+    _Stop_callback_base(const _Stop_callback_base&) = delete;
+    _Stop_callback_base(_Stop_callback_base&&)      = delete;
+    _Stop_callback_base& operator=(const _Stop_callback_base&) = delete;
+    _Stop_callback_base& operator=(_Stop_callback_base&&) = delete;
+
+    // if _Token is _Stop_requested, calls the callback;
+    // otherwise, inserts *this into the callback list if stop is possible
+    inline void _Attach(const stop_token& _Token) noexcept;
+    inline void _Attach(stop_token&& _Token) noexcept;
+
+    // if *this is in a callback list, removes it
+    inline void _Detach() noexcept;
+
+private:
+    template <bool _Transfer_ownership>
+    void _Do_attach(conditional_t<_Transfer_ownership, _Stop_state*&, _Stop_state* const> _State) noexcept;
+
+protected:
+    _Stop_state* _Parent       = nullptr;
+    _Stop_callback_base* _Next = nullptr;
+    _Stop_callback_base* _Prev = nullptr;
+    _Callback_fn _Fn;
+};
+
+struct _Stop_state {
+    atomic<uint32_t> _Stop_tokens  = 1; // plus one shared by all stop_sources
+    atomic<uint32_t> _Stop_sources = 2; // plus the low order bit is the stop requested bit
+    _Locked_pointer<_Stop_callback_base> _Callbacks;
+    // always uses relaxed operations; ordering provided by the _Callbacks lock
+    // (atomic just to get wait/notify support)
+    atomic<const _Stop_callback_base*> _Current_callback = nullptr;
+    _Thrd_id_t _Stopping_thread                          = 0;
+
+    [[nodiscard]] bool _Stop_requested() const noexcept {
+        return (_Stop_sources.load() & uint32_t{1}) != 0;
+    }
+
+    [[nodiscard]] bool _Stop_possible() const noexcept {
+        return _Stop_sources.load() != 0;
+    }
+
+    [[nodiscard]] bool _Request_stop() noexcept {
+        // Attempts to request stop and call callbacks, returns whether request was successful
+        if ((_Stop_sources.fetch_or(uint32_t{1}) & uint32_t{1}) != 0) {
+            // another thread already requested
+            return false;
+        }
+
+        _Stopping_thread = _Thrd_id();
+        for (;;) {
+            auto _Head = _Callbacks._Lock_and_load();
+            _Current_callback.store(_Head, memory_order_relaxed);
+            _Current_callback.notify_one();

```

Comment:
I suspect this might need to be `notify_all()` to handle the case where two threads are concurrently waiting to deregister callbacks. The semantics of `atomic::notify_one()` is a bit of a dark corner that I'd be wary of relying on here.

Reply:
I think `notify_one` is fine: there can be at most one thread that waits for `_Current_callback` to not be `_Head`, the one calling the dtor of that thing; but I also don't think it really matters so I changed it to notify_all.